subprojects/gdk_rust/gdk_electrum/src/lib.rs

```diff
@@ -0,0 +1,1212 @@
+                            }
+
+                            db_for_registry.insert_asset_registry(&registry).unwrap();
+                            db_for_registry.insert_asset_icons(&icons).unwrap();
+                        }
+                        _ => warn!("Registry or icons are not json"),
+                    }
+                } else {
+                    warn!("Cannot download registry and icons");
+                }
+            }))
+        } else {
+            None
+        };
+
+        let checker = match self.network.id() {
+            NetworkId::Bitcoin(network) => {
+                let mut path: PathBuf = self.data_root.as_str().into();
+                path.push(format!("headers_chain_{}", network));
+                ChainOrVerifier::Chain(HeadersChain::new(path, network)?)
+            }
+            NetworkId::Elements(network) => {
+                let verifier = Verifier::new(network);
+                ChainOrVerifier::Verifier(verifier)
+            }
+        };
+
+        let mut headers_kind = match &self.url {
+            ElectrumUrl::Tls(url, validate) => {
+                let client = electrum_client::Client::new_ssl(url.as_str(), *validate)?;
+                HeadersKind::Tls(
+                    Headers::new(db.clone(), checker, client)?,
+                    url.to_string(),
+                    *validate,
+                )
+            }
+            ElectrumUrl::Plaintext(url) => {
+                let client = electrum_client::Client::new(&url)?;
+                HeadersKind::Plain(Headers::new(db.clone(), checker, client)?, url.to_string())
+            }
+        };
+        let (close_headers, r) = channel();
+        self.closer.senders.push(close_headers);
+        let mut chunk_size = DIFFCHANGE_INTERVAL as usize;
+        let headers_handle = thread::spawn(move || {
+            info!("starting headers thread");
+            'outer: loop {
+                if wait_or_close(&r, sync_interval) {
+                    info!("closing headers thread");
+                    break;
+                }
+
+                loop {
+                    match headers_kind.ask(chunk_size) {
+                        Ok(headers_found) => {
+                            if headers_found == 0 {
+                                chunk_size = 1
+                            } else {
+                                info!("headers found: {}", headers_found);
+                            }
+                        }
+                        Err(Error::InvalidHeaders) => {
+                            // this should handle reorgs and also broke IO writes update
+                            if headers_kind.remove(144).is_err() {
+                                break;
+                            }
+                        }
+                        Err(e) => {
+                            // usual error is because I reached the tip, trying asking half
```

Comment:
This should not result in an error. From the [ElectrumX docs](https://electrumx.readthedocs.io/en/latest/protocol-methods.html#blockchain-block-headers):

> If the chain has not extended sufficiently far, only the available headers will be returned. If more headers than max were requested at most max will be returned.

I [changed this in bwt](https://github.com/shesek/bwt/commit/fb0c4b517eec6d6f1c69079da25115f1d8d8a3f6) awhile back, but I checked now and it appears that esplora does incorrectly return an error. I added myself a TODO to get this fixed. (I also checked upstream electrs, and it does implement this correctly.)

Reply:
added a comment to remember to update in 20d4640f189d4fffabfc3d7e7fa3148673a74947

Reply:
Fixed in https://github.com/Blockstream/electrs/commit/ff9237df4223468d4f1260cc377a013cf3a1ed25.